cdap-ui/app/cdap/components/FastAction/SetPreferenceAction/SetPreferenceModal.js

```diff
@@ -0,0 +1,430 @@
+    const keyLabel = T.translate('features.FastAction.setPreferencesColumnLabel.key');
+    const valueLabel = T.translate('features.FastAction.setPreferencesColumnLabel.value');
+    let description = T.translate('features.FastAction.setPreferencesDescriptionLabel.namespace');
+    if (this.props.entity) {
+      if (this.props.entity.type == 'application') {
+        description = T.translate('features.FastAction.setPreferencesDescriptionLabel.app');
+      } else {
+        description = T.translate('features.FastAction.setPreferencesDescriptionLabel.program');
+      }
+    }
+    return (
+      <div>
+        <div className='specify-preferences-description'>
+          <h4>{title}</h4>
+          <p>{description}</p>
+        </div>
+        <div className='specify-preferences-list'>
+          <div className='specify-preferences-labels'>
+            <span className='key-label'>{keyLabel}</span>
+            <span className='value-label'>{valueLabel}</span>
+          </div>
+          <div className='specify-preferences-values'>
+            <KeyValuePairs
+              keyValues = {this.state.keyValues}
+              onKeyValueChange = {this.onKeyValueChange}
+              fieldsResetted = {this.state.fieldsResetted}
+            />
+          </div>
+        </div>
+      </div>
+    );
+  }
+
+  renderInheritedPreferencesColumnHeader(column) {
+    let columnToAttribute = column.toLowerCase();
+    return (
+      <th>
+        <span
+          className='toggleable-columns'
+          onClick={this.toggleSorted.bind(this, columnToAttribute)}
+        >
+          {
+            this.state.sortByAttribute === columnToAttribute ?
+              <span>
+                <span className='text-underline'>{column}</span>
+                <span>
+                  {
+                    this.state.sortOrder === 'asc' ?
+                      <i className='fa fa-caret-down fa-lg'></i>
+                    :
+                      <i className='fa fa-caret-up fa-lg'></i>
+                  }
+                </span>
+              </span>
+            :
+              <span>{column}</span>
+          }
+        </span>
+      </th>
+    );
+  }
+
+  renderInheritedPreferences() {
+    const titleLabel = T.translate('features.FastAction.setPreferencesInheritedPrefsLabel');
+    const keyLabel = T.translate('features.FastAction.setPreferencesColumnLabel.key');
+    const valueLabel = T.translate('features.FastAction.setPreferencesColumnLabel.value');
+    const originLabel = T.translate('features.FastAction.setPreferencesColumnLabel.origin');
+    let numInheritedPreferences = this.state.inheritedPreferences.length;
+    return (
+      <div>
+        <div className='inherited-preferences-label'>
+          <h4>{titleLabel} ({numInheritedPreferences})</h4>
+        </div>
+        <div className='inherited-preferences-list'>
+        {
+           numInheritedPreferences ?
+            <div>
+              <table>
```

Comment:
Shall we do this in a different PR? I think it'd take some time to extract out the component, and still make sure the styling looks the same. Just not sure it'll be completed if we're trying to merge this PR before today's noon/evening.